File: ios/Classes/lib/notifications/NotificationModel.swift
//
//  NotificationModel.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 05/09/20.
//

import Foundation

public class NotificationModel : AbstractModel {
        
    var content:NotificationContentModel?
    var actionButtons:[NotificationButtonModel]?
    var schedule:NotificationScheduleModel?
    
    public func fromMap(arguments: [String : Any?]?) -> AbstractModel? {
        
        do {
            self.content = extractNotificationContent(Definitions.NOTIFICATION_MODEL_CONTENT, arguments)
            
            // required
            if(self.content == nil){ return nil }
            
            self.schedule = try extractNotificationSchedule(Definitions.NOTIFICATION_MODEL_SCHEDULE, arguments)
            self.actionButtons = extractNotificationButtons(Definitions.NOTIFICATION_MODEL_BUTTONS, arguments)
            
            return self
        
        }
        catch {
            print("error " + error.localizedDescription)
        }
            
        return nil
    }
    
    public func toMap() -> [String : Any?] {
        var mapData:[String: Any?] = [:]
        
        mapData["content"] = self.content!.toMap()
        if(self.schedule != nil){ mapData["schedule"] = self.schedule!.toMap() }
        if(self.actionButtons != nil){
            var listButtons:[[String:Any?]] = []
            
            for button in self.actionButtons! {
                listButtons.append(button.toMap())
            }
            
            mapData["actionButtons"] = listButtons
        }
        
        return mapData
    }
    
    func extractNotificationContent(_ reference:String, _ arguments:[String:Any?]?) -> NotificationContentModel? {
        guard let map:[String:Any?] = arguments?[reference] as? [String:Any?] else { return nil }
        if(map.isEmpty){ return nil }
        return NotificationContentModel().fromMap(arguments: map) as? NotificationContentModel
    }
    
    func extractNotificationSchedule(_ reference:String, _ arguments:[String:Any?]?) throws -> NotificationScheduleModel? {
        guard let map:[String:Any?] = arguments?[reference] as? [String:Any?] else { return nil }
        if(map.isEmpty){ return nil }
        
        if(
            map[Definitions.NOTIFICATION_CRONTAB_EXPRESSION] != nil ||
            map[Definitions.NOTIFICATION_PRECISE_SCHEDULES] != nil ||
            map[Definitions.NOTIFICATION_INITIAL_DATE_TIME] != nil ||
            map[Definitions.NOTIFICATION_EXPIRATION_DATE_TIME] != nil
        ){
            throw AwesomeNotificationsException.invalidRequiredFields(msg: "Crontab schedules are not available for iOS")
        }
        
        if map["interval"] != nil {
            return NotificationIntervalModel().fromMap(arguments: map) as? NotificationScheduleModel
        }
        else {
            return NotificationCalendarModel().fromMap(arguments: map) as? NotificationScheduleModel
        }
    }
    
    func extractNotificationButtons(_ reference:String, _ arguments:[String:Any?]?) -> [NotificationButtonModel]? {
        guard let actionButtonsData:[[String:Any?]] = arguments?[reference] as? [[String:Any?]] else { return nil }
        if(actionButtonsData.isEmpty){ return nil }
        
        var actionButtons:[NotificationButtonModel] = []
        
        for buttonData in actionButtonsData {
            let button:NotificationButtonModel? = NotificationButtonModel().fromMap(arguments: buttonData) as? NotificationButtonModel
            if(button == nil){ return nil }
            actionButtons.append(button!)
        }
        
        return actionButtons
    }
    
    public func validate() throws {
        try self.content?.validate()
        try self.schedule?.validate()
        
        if(self.actionButtons != nil){
            for button in self.actionButtons! {
                try button.validate()
            }
        }
    }
}

File: example/android/android_core/src/main/java/me/carda/awesome_notifications/core/models/NotificationModel.java
package me.carda.awesome_notifications.core.models;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import me.carda.awesome_notifications.core.Definitions;
import me.carda.awesome_notifications.core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.core.exceptions.ExceptionCode;
import me.carda.awesome_notifications.core.exceptions.ExceptionFactory;

public class NotificationModel extends AbstractModel {

    private static final String TAG = "NotificationModel";

    public boolean groupSummary = false;
    public String remoteHistory;

    public NotificationContentModel content;
    public NotificationScheduleModel schedule;
    public List<NotificationButtonModel> actionButtons;

    public NotificationModel(){}

    public NotificationModel ClonePush(){
        return new NotificationModel().fromMap(this.toMap());
    }

    @Override
    @Nullable
    public NotificationModel fromMap(Map<String, Object> parameters){
        content = extractNotificationContent(parameters);
        if(content == null) return null;

        schedule = extractNotificationSchedule(parameters);
        actionButtons = extractNotificationButtons(parameters);

        return this;
    }

    @Override
    public Map<String, Object> toMap(){
        if(content == null) return null;
        Map<String, Object> dataMap = new HashMap<>();

        putDataOnSerializedMap(Definitions.NOTIFICATION_MODEL_CONTENT, dataMap, content);
        putDataOnSerializedMap(Definitions.NOTIFICATION_MODEL_SCHEDULE, dataMap, schedule);
        putDataOnSerializedMap(Definitions.NOTIFICATION_MODEL_BUTTONS, dataMap, actionButtons);

        return dataMap;
    }

    @Override
    public String toJson() {
        return templateToJson();
    }

    @Override
    public NotificationModel fromJson(String json){
        return (NotificationModel) super.templateFromJson(json);
    }

    private static NotificationContentModel extractNotificationContent(
            @NonNull Map<String, Object> parameters
    ){
        if(!parameters.containsKey(Definitions.NOTIFICATION_MODEL_CONTENT)) return null;
        Object obj = parameters.get(Definitions.NOTIFICATION_MODEL_CONTENT);

        if(!(obj instanceof Map<?,?>)) return null;

        @SuppressWarnings("unchecked")
        Map<String, Object> map = (Map<String, Object>) obj;

        if(map.isEmpty()) return null;
        else return new NotificationContentModel().fromMap(map);
    }

    private static NotificationScheduleModel extractNotificationSchedule(
            @NonNull Map<String, Object> parameters
    ){
        if(!parameters.containsKey(Definitions.NOTIFICATION_MODEL_SCHEDULE)) return null;
        Object obj = parameters.get(Definitions.NOTIFICATION_MODEL_SCHEDULE);

        if(!(obj instanceof Map<?,?>)) return null;

        @SuppressWarnings("unchecked")
        Map<String, Object> map = (Map<String, Object>) obj;

        return NotificationScheduleModel.getScheduleModelFromMap(map);
    }

    @SuppressWarnings("unchecked")
    private static List<NotificationButtonModel> extractNotificationButtons(
            @NonNull Map<String, Object> parameters
    ){
        if(!parameters.containsKey(Definitions.NOTIFICATION_MODEL_BUTTONS)) return null;
        Object obj = parameters.get(Definitions.NOTIFICATION_MODEL_BUTTONS);

        if(!(obj instanceof List<?>)) return null;
        List<Object> actionButtonsData = (List<Object>) obj;

        List<NotificationButtonModel> actionButtons = new ArrayList<>();

        for (Object objButton: actionButtonsData) {
            if(!(objButton instanceof Map<?,?>)) return null;

            Map<String, Object> map = (Map<String, Object>) objButton;
            if(map.isEmpty()) continue;

            NotificationButtonModel button = new NotificationButtonModel().fromMap(map);
            actionButtons.add(button);
        }

        if(actionButtons.isEmpty()) return null;

        return actionButtons;
    }

    public void validate(
            Context context
    ) throws AwesomeNotificationsException {
        if(this.content == null)
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Notification content is required",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationContent");

        this.content.validate(context);

        if(this.schedule != null)
            this.schedule.validate(context);

        if(this.actionButtons != null){
            for(NotificationButtonModel button : this.actionButtons){
                button.validate(context);
            }
        }
    }

}
